src/triage/component/catwalk/__init__.py

```diff
@@ -104,34 +108,63 @@ def process_task(self, test_store, train_store, train_kwargs):
 
             # Generate predictions for the testing data then training data
             for store in (test_store, train_store):
-                if self.predictor.replace or self.model_evaluator.needs_evaluations(store, model_id):
+
+                if self.predictor.replace:
                     logging.info(
-                        "The evaluations needed for matrix %s-%s and model %s"
-                        "are not all present in db, so predicting and evaluating",
+                        "Replace flag set; generating new predictions and evaluations for"
+                        "matrix %s-%s, and model %s",
                         store.uuid,
                         store.matrix_type,
                         model_id
                     )
+
                     predictions_proba = self.predictor.predict(
                         model_id,
                         store,
                         misc_db_parameters=dict(),
                         train_matrix_columns=train_store.columns(),
                     )
 
-                    self.model_evaluator.evaluate(
-                        predictions_proba=predictions_proba,
-                        matrix_store=store,
-                        model_id=model_id,
-                    )
+                    for subset in self.subsets:
+                        self.model_evaluator.evaluate(
+                            predictions_proba=predictions_proba,
+                            matrix_store=store,
+                            model_id=model_id,
+                            subset=subset,
+                        )
+                    
```

Comment:
I did find this splitup of code surprising at first read so let's unpack.

Before, we had one 'if/else' block before to either predict/evaluate or not, that took into account both predictions.

Now, it's split up. Just based on the replace flag, we either re-run everything, or we go look per-subset to see if any are needed.

One problem I see after looking at this further: in the else block, where do the predictions_proba come from? It doesn't look like we're calling predict in that case. It looks like this else block exists to allow us to only evaluate subsets if needed; if we were fine with re-doing all subset evaluations in the event that *any* subset evaluations were needed, we could do something more similar to what the old code was.

If conditionally doing subset evaluations is desirable for time-saving purposes, then I think we would have to do something a bit different to avoid unnecessarily doing prediction (which does spend a bunch of time on I/O) when we don't need to. But I want to make sure this is the case before we do that. In my experience with evaluation time before, it's that *the evaluation itself* is very quick once we have a `predictions_proba`, it's that we want to make very sure that we need a `predictions_proba` before we go and create it. That's why both the predict and evaluate are under the same conditional in the old code, so that we can avoid doing prediction if no evaluations are needed.

Reply:
Hmmm, good points. Let's chat about this. I was thinking of my previous experience with very slow subset queries and wanted to check for each subset whether it really needs to be run, but perhaps that's not a great solution -- if a subset query is slow, the predictions will be painful to write the first time, too, which would hopefully spur some optimization on the user end.